group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1867,11 +1936,11 @@ private Assignment updateTargetAssignment(
                     .withTopicsImage(metadataImage.topics())
                     .addOrUpdateMember(updatedMember.memberId(), updatedMember);
 
-            if (staticMemberReplaced) {
-                // A new static member is replacing an older one with the same subscriptions.
-                // We just need to remove the older member and add the newer one. The new member should
-                // reuse the target assignment of the older member.
-                assignmentResultBuilder.removeMember(member.memberId());
+            // If the instance id was associated to a different member, it means that the
+            // static member is replaced by the current member hence we remove the previous one.
+            String previousMemberId = group.staticMemberId(updatedMember.instanceId());
+            if (!updatedMember.memberId().equals(previousMemberId)) {
+                assignmentResultBuilder.removeMember(previousMemberId);
```

Comment:
`previousMemberId` could be null now, so `assignmentResultBuilder` will put (null, null) into `updatedMembers`. `HashMap` can handle null key well but it seems to me that is a bit weird. WDYT?